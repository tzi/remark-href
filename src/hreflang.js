import select from "unist-util-select";

const pattern = new RegExp(" \\(([a-z]{2}(_[a-z]{2}|))\\)$", "i");

function plugin() {
  return function transformer(ast) {
    select(ast, "link,linkReference").forEach(node => {
      if (!node.children) {
        return;
      }
      const lastChild = node.children[node.children.length - 1];
      const match = pattern.exec(lastChild.value);

      if (match) {
        lastChild.value = lastChild.value.slice(0, -match[0].length);
        node.data = {
          hProperties: {
            hreflang: match[1]
          }
        };
      }
    });
  };
}

module.exports = plugin;
