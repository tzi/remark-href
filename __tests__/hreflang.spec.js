import dedent from "dedent";
import unified from "unified";
import reParse from "remark-parse";
import stringify from "rehype-stringify";
import remark2rehype from "remark-rehype";

import hreflang from "../src/hreflang";

const render = text =>
  unified()
    .use(reParse)
    .use(hreflang)
    .use(remark2rehype)
    .use(stringify)
    .processSync(text);

describe("hreflang", function() {
  it("renders a link with a two digits lang", () => {
    const { contents } = render(dedent`
      I use this plugin on [my own website (fr)](https://tzi.fr), because I mix several languages on it!
    `);

    expect(contents).toMatchSnapshot();
  });

  it("renders a link with a four digits lang", () => {
    const { contents } = render(dedent`
      I love [Canada (fr_ca)](https://www.canada.ca/fr.html)!
    `);

    expect(contents).toMatchSnapshot();
  });

  it("renders a link with an invalid lang format", () => {
    const { contents } = render(dedent`
      The [Kaamelott TV shows (Celtes)](https://www.6play.fr/kaamelott-p_888) is really fun. 
    `);

    expect(contents).toMatchSnapshot();
  });

  it("renders a link with a reference for URL", () => {
    const {contents} = render(dedent`
      Did you get this [reference (fr)][reference]?

      [reference]: https://example.com
    `);

    expect(contents).toMatchSnapshot();
  });
});
