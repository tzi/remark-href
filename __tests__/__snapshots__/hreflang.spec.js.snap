// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hreflang renders a link with a four digits lang 1`] = `"<p>I love <a href=\\"https://www.canada.ca/fr.html\\" hreflang=\\"fr_ca\\">Canada</a>!</p>"`;

exports[`hreflang renders a link with a reference for URL 1`] = `"<p>Did you get this <a href=\\"https://example.com\\" hreflang=\\"fr\\">reference</a>?</p>"`;

exports[`hreflang renders a link with a two digits lang 1`] = `"<p>I use this plugin on <a href=\\"https://tzi.fr\\" hreflang=\\"fr\\">my own website</a>, because I mix several languages on it!</p>"`;

exports[`hreflang renders a link with an invalid lang format 1`] = `"<p>The <a href=\\"https://www.6play.fr/kaamelott-p_888\\">Kaamelott TV shows (Celtes)</a> is really fun.</p>"`;
